src/entity_builder.rs

```diff
@@ -53,38 +53,98 @@ impl EntityBuilder {
     pub fn add<T: Component>(&mut self, component: T) -> &mut Self {
-        if !self.id_set.insert(TypeId::of::<T>()) {
-            return self;
-        }
-        let end = self.cursor + mem::size_of::<T>();
-        if end > self.storage.len() {
-            self.grow(end);
-        }
-        if mem::size_of::<T>() != 0 {
-            unsafe {
-                self.storage
-                    .as_mut_ptr()
-                    .add(self.cursor)
-                    .cast::<T>()
-                    .write_unaligned(component);
+        match self.indices.entry(TypeId::of::<T>()) {
+            Entry::Occupied(occupied) => {
+                let index = *occupied.get();
+                let (_, offset) = self.info[index];
+                unsafe {
+                    let storage_ptr = self
+                        .storage
+                        .as_mut_ptr()
+                        .cast::<u8>()
+                        .add(offset)
+                        .cast::<T>();
+
+                    // alloc a properly aligned tmp buffer and copy in the old value
+                    // so we can drop it safely
+                    let _ = storage_ptr.read_unaligned();
+                    // Overwrite the old value with our new one.
+                    storage_ptr.write_unaligned(component);
+                }
+                self
             }
+            Entry::Vacant(vacant) => {
+                let end = self.cursor + mem::size_of::<T>();
+                if end > self.storage.len() {
+                    Self::grow(end, self.cursor, &mut self.storage);
+                }
+                unsafe {
+                    self.storage
+                        .as_mut_ptr()
+                        .add(self.cursor)
+                        .cast::<T>()
+                        .write_unaligned(component);
+                }
+                vacant.insert(self.info.len());
+                self.info.push((TypeInfo::of::<T>(), self.cursor));
+                self.cursor += mem::size_of::<T>();
+                self
+            }
+        }
+    }
+
+    /// Add all components in `bundle` to the entity
+    pub fn add_bundle(&mut self, bundle: impl DynamicBundle) -> &mut Self {
+        unsafe {
+            bundle.put(|ptr, ty| {
+                match self.indices.entry(ty.id()) {
+                    Entry::Occupied(occupied) => {
+                        let index = *occupied.get();
+                        let (ty, offset) = self.info[index];
+
+                        let storage_ptr = self.storage.as_mut_ptr().cast::<u8>().add(offset);
+                        // alloc a properly aligned tmp buffer and copy in the old value
+                        // so we can drop it safely
+                        let tmp = alloc(ty.layout());
+                        ptr::copy_nonoverlapping(storage_ptr, tmp, ty.layout().size());
+                        ty.drop(tmp);
+                        dealloc(tmp, ty.layout());
```

Comment:
The fiddly dance here makes me think we should probably just store things properly aligned in the first place. Then I think there'd be no reason for `add` not to dispatch directly to `add_bundle`, which saves a lot of duplication. If you prefer we can leave that for future work, but I think it'd be a substantial benefit.

Reply:
It shouldn't be difficult to do that and I agree that it would be good for `add` to dispatch directly to `add_bundle`. I'll look at making that improvement but I think it's substantial enough for me that I'd like to get this PR in first before I work on it.